packages/fetchai/skills/tac_control_contract/handlers.py

```diff
@@ -0,0 +1,427 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""This package contains the handlers."""
+
+from typing import Optional, cast
+
+from aea.configurations.base import ProtocolId
+from aea.crypto.base import LedgerApi
+from aea.decision_maker.messages.transaction import TransactionMessage
+from aea.protocols.base import Message
+from aea.skills.base import Handler
+
+from packages.fetchai.protocols.oef.message import OEFMessage
+from packages.fetchai.protocols.tac.message import TACMessage
+from packages.fetchai.protocols.tac.serialization import TACSerializer
+from packages.fetchai.skills.tac_control_contract.game import Game, Phase, Transaction
+from packages.fetchai.skills.tac_control_contract.parameters import Parameters
+
+
+class TACHandler(Handler):
+    """This class handles oef messages."""
+
+    SUPPORTED_PROTOCOL = TACMessage.protocol_id
+
+    def setup(self) -> None:
+        """
+        Implement the handler setup.
+
+        :return: None
+        """
+        pass
+
+    def handle(self, message: Message) -> None:
+        """
+        Handle a register message.
+
+        If the address is already registered, answer with an error message.
+
+        :param message: the 'get agent state' TACMessage.
+        :return: None
+        """
+        tac_message = cast(TACMessage, message)
+        tac_type = tac_message.type
+
+        game = cast(Game, self.context.game)
+
+        self.context.logger.debug(
+            "[{}]: Handling TAC message. type={}".format(
+                self.context.agent_name, tac_type
+            )
+        )
+        if (
+            tac_type == TACMessage.Type.REGISTER
+            and game.phase == Phase.GAME_REGISTRATION
+        ):
+            self._on_register(tac_message)
+        elif (
+            tac_type == TACMessage.Type.UNREGISTER
+            and game.phase == Phase.GAME_REGISTRATION
+        ):
+            self._on_unregister(tac_message)
+        elif tac_type == TACMessage.Type.TRANSACTION and game.phase == Phase.GAME:
```

Comment:
Note, we no longer will accept the `Transaction` as a message type in this handler.